File: python/examples/image_blur_fft.py
#!/usr/bin/env python3
"""
Image Blur via FFT Convolution

Demonstrates:
- Creating Gaussian kernels
- FFT-based convolution (faster for large kernels)
- Comparing blur levels
- Analyzing frequency content changes
"""

import numpy as np
import spectrograms as sg

def main():
    print("=== Image Blur via FFT Convolution ===\n")

    # Create a test image with sharp features
    size = 256
    print(f"Image size: {size} x {size}")

    # Create image with sharp edges and patterns
    image = np.zeros((size, size))

    # Add some rectangles
    image[50:100, 50:100] = 1.0
    image[150:200, 100:200] = 0.8

    # Add a checkerboard pattern
    for i in range(0, size, 32):
        for j in range(0, size, 32):
            if (i + j) % 64 == 0:
                image[i:i+16, j:j+16] = 0.5

    print(f"Original image range: [{image.min():.2f}, {image.max():.2f}]")
    print(f"Original image mean: {image.mean():.3f}\n")

    # === 1. Create Gaussian Kernels ===
    print("1. Creating Gaussian blur kernels...")

    blur_levels = [
        (5, 1.0, "Light"),
        (9, 2.0, "Medium"),
        (15, 3.0, "Heavy"),
    ]

    kernels = []
    for size_k, sigma, name in blur_levels:
        kernel = sg.gaussian_kernel_2d(size_k, sigma)
        kernels.append((kernel, name))
        print(f"   {name} blur: kernel size {size_k}x{size_k}, sigma={sigma}")
        print(f"     Kernel sum: {kernel.sum():.6f} (should be ~1.0)")
        print(f"     Kernel max: {kernel.max():.6f}")

    print()

    # === 2. Apply Blurs via FFT Convolution ===
    print("2. Applying blurs via FFT convolution...")

    blurred_images = []
    for kernel, name in kernels:
        blurred = sg.convolve_fft(image, kernel)
        blurred_images.append((blurred, name))

        print(f"   {name} blur:")
        print(f"     Output range: [{blurred.min():.3f}, {blurred.max():.3f}]")
        print(f"     Output mean: {blurred.mean():.3f}")

    print()

    # === 3. Analyze Frequency Content ===
    print("3. Analyzing frequency content changes...")

    # Original image spectrum
    original_power = sg.power_spectrum_2d(image)
    original_total_power = original_power.sum()

    print(f"   Original image:")
    print(f"     Total power: {original_total_power:.2e}")
    print(f"     DC component: {original_power[0, 0]:.2e}")

    # Blurred images spectra
    for blurred, name in blurred_images:
        blurred_power = sg.power_spectrum_2d(blurred)
        blurred_total_power = blurred_power.sum()

        # Compare power in different frequency bands
        # Low frequencies (DC + nearby)
        low_freq_power = blurred_power[:5, :3].sum()
        high_freq_power = blurred_power[20:, 20:].sum() if blurred_power.shape[0] > 20 else 0

        print(f"\n   {name} blur:")
        print(f"     Total power: {blurred_total_power:.2e}")
        print(f"     Power ratio: {blurred_total_power / original_total_power:.3f}")
        print(f"     Low freq power: {low_freq_power:.2e}")
        print(f"     High freq power: {high_freq_power:.2e}")

    print()

    # === 4. Compare Statistics ===
    print("4. Comparing blur statistics...")

    print(f"\n   {'Blur Level':<15} {'Min':<10} {'Max':<10} {'Mean':<10} {'Std Dev':<10}")
    print(f"   {'-'*60}")

    # Original
    print(f"   {'Original':<15} {image.min():<10.3f} {image.max():<10.3f} "
          f"{image.mean():<10.3f} {image.std():<10.3f}")

    # Blurred versions
    for blurred, name in blurred_images:
        print(f"   {name + ' blur':<15} {blurred.min():<10.3f} {blurred.max():<10.3f} "
              f"{blurred.mean():<10.3f} {blurred.std():<10.3f}")

    print()

    # === 5. Direct Comparison ===
    print("5. Effect of blur on sharp edges...")

    # Sample a line through a sharp edge
    row = 75  # Middle of first rectangle
    original_line = image[row, 40:110]
    print(f"\n   Original edge profile (row {row}, cols 40-110):")
    print(f"     Max gradient: {np.abs(np.diff(original_line)).max():.3f}")

    for blurred, name in blurred_images:
        blurred_line = blurred[row, 40:110]
        max_gradient = np.abs(np.diff(blurred_line)).max()
        print(f"   {name} blur:")
        print(f"     Max gradient: {max_gradient:.3f} (edge smoothness)")

    print("\n=== Example Complete ===")
    print("\nKey observations:")
    print("- FFT convolution preserves total image energy")
    print("- Heavier blurs reduce high-frequency content")
    print("- Sharp edges become smoother with increased blur")
    print("- Kernel sum = 1.0 ensures brightness preservation")
    print("\nFFT convolution is faster than spatial convolution for kernels > 7x7")


if __name__ == "__main__":
    main()

File: examples/image_blur_fft.rs
//! Gaussian blur using FFT convolution
//!
//! Demonstrates applying Gaussian blur to an image using FFT-based convolution,
//! which is faster than spatial convolution for large kernels.

use std::num::NonZeroUsize;

use ndarray::Array2;
use spectrograms::image_ops::{convolve_fft, gaussian_kernel_2d};
use spectrograms::{SpectrogramResult, nzu};

fn main() -> SpectrogramResult<()> {
    println!("=== Gaussian Blur via FFT Convolution ===\n");

    // Create a test image with a bright square in the center
    let size = 128;
    let image = Array2::<f64>::from_shape_fn((size, size), |(i, j)| {
        let center_i = size as f64 / 2.0;
        let center_j = size as f64 / 2.0;

        // Create a 40x40 square in the center
        if (i as f64 - center_i).abs() < 20.0 && (j as f64 - center_j).abs() < 20.0 {
            100.0 // Bright square
        } else {
            0.0 // Dark background
        }
    });

    println!("Created {}x{} test image with bright square", size, size);

    // Compute statistics on original image
    let original_mean: f64 = image.iter().sum::<f64>() / (size * size) as f64;
    let original_min = image.iter().cloned().fold(f64::INFINITY, f64::min);
    let original_max = image.iter().cloned().fold(f64::NEG_INFINITY, f64::max);

    println!("\nOriginal image statistics:");
    println!(
        "  Min: {:.2}, Max: {:.2}, Mean: {:.2}",
        original_min, original_max, original_mean
    );

    // Apply different levels of blur
    let blur_configs = vec![(5, 1.0, "Light"), (9, 2.0, "Medium"), (15, 3.0, "Heavy")];

    for (kernel_size, sigma, description) in blur_configs {
        println!(
            "\n{} blur (kernel size: {}, sigma: {}):",
            description, kernel_size, sigma
        );

        // Create Gaussian kernel
        let kernel_size = NonZeroUsize::new(kernel_size).unwrap();
        let kernel = gaussian_kernel_2d(kernel_size, sigma)?;

        // Verify kernel is normalized
        let kernel_sum: f64 = kernel.iter().sum();
        println!("  Kernel sum: {:.6} (should be 1.0)", kernel_sum);

        // Apply blur via FFT convolution
        let blurred = convolve_fft(&image.view(), &kernel.view())?;

        // Compute statistics on blurred image
        let blurred_mean: f64 = blurred.iter().sum::<f64>() / (size * size) as f64;
        let blurred_min = blurred.iter().cloned().fold(f64::INFINITY, f64::min);
        let blurred_max = blurred.iter().cloned().fold(f64::NEG_INFINITY, f64::max);

        println!("  Blurred image:");
        println!(
            "    Min: {:.2}, Max: {:.2}, Mean: {:.2}",
            blurred_min, blurred_max, blurred_mean
        );
        println!(
            "    Peak reduction: {:.1}%",
            (original_max - blurred_max) / original_max * 100.0
        );
    }

    // Compare with edge statistics
    println!("\n=== Edge Analysis ===\n");

    let light_kernel = gaussian_kernel_2d(nzu!(5), 1.0)?;
    let blurred = convolve_fft(&image.view(), &light_kernel.view())?;

    // Sample values at center and edge
    let center_val = blurred[[size / 2, size / 2]];
    let edge_val = blurred[[size / 2, 0]];

    println!("Sample values:");
    println!("  Center (inside square): {:.2}", center_val);
    println!("  Edge (outside square): {:.2}", edge_val);
    println!("  Ratio: {:.1}:1", center_val / edge_val.max(1e-10));

    println!("\nExample completed successfully!");

    Ok(())
}
